lib/phoenix_live_view/channel.ex

```diff
@@ -465,6 +486,25 @@ defmodule Phoenix.LiveView.Channel do
     end
   end
 
+  defp unregister_upload(state, ref, entry_ref, cid) do
+    {new_state, %UploadConfig{} = conf} =
+      with_socket(state, cid, nil, fn socket, _ ->
```

Comment:
Remember `with_socket` can fail due to race conditions between processing an event and the component being removed from the page, so you need to consider this potentially returning `{state, :error}`.

However, I think there is a simpler solution. All uses of `with_socket`, except for `fetch_upload_config`, only want to return something from `with_socket` to change the state! So instead, we can return the updated state from `with_socket` and skip this indirection. For example, this function would do:

```elixir
  defp unregister_upload(state, ref, entry_ref, cid) do
    with_socket(state, cid, nil, fn socket, _ ->
      conf = Upload.get_upload_by_ref!(socket, ref)
      {_, new_state} = pop_in(state.upload_names[conf.name])
      {Upload.unregister_completed_entry_upload(socket, conf, entry_ref), {:ok, nil, new_state}}
    end)
  end
```

Look how we don't need to touch the state after. And `with_socket` would do:

```elixir
  # If :error is returned, the socket must not change,
  # otherwise we need to call push_diff on all cases.
  defp with_socket(state, nil, ref, fun) do
    {new_socket, return} = fun.(state.socket, nil)

    case return do
      {:ok, ref_reply, state} ->
        {:noreply, new_state} = handle_changed(state, new_socket, ref_reply)
        new_state

      :error ->
        push_noop(state, ref)
    end
  end

  defp with_socket(state, cid, ref, fun) do
    %{socket: socket, components: components} = state

    {diff, new_components, return} =
      case Diff.with_component(socket, cid, %{}, components, fun) do
        {diff, new_components, return} = triplet ->
          triplet

        :error ->
          {%{}, components, :error}
      end

    case return do
      {:ok, ref_reply, state} ->
        new_state = %{state | components: new_components}
        push_diff(new_state, diff, ref_reply)

      :error ->
        push_noop(new_state, ref)
    end
  end
```

This should clean up all `with_socket` callers, except `fetch_upload_config`, which I think we should tackle differently (i.e. we shouldn't even call Diff.with_component because we are only interested in reading a socket but not really change it. I think we need a `Diff.read_component` function (and rename `with_component` to `write_component`).